File: src/Council.Protocol/Loading/ProtocolLoader.cs
using System.Security;
using System.Text.Json;
using Council.Protocol.Diagnostics;

namespace Council.Protocol.Loading;

/// <summary>
/// Protocol 文件 → 读取 → JSON 解析 → 原始 Protocol / 解析错误。
/// <para>
/// 职责边界（D1-R1 §5.1）：本类型<b>不</b>负责 Schema 校验、语义校验、
/// Meeting Runtime、Agent 调度、UI、Prompt 或 Transport。
/// </para>
/// </summary>
public sealed class ProtocolLoader
{
    private const string SearchPattern = "*.json";

    private static readonly JsonDocumentOptions ParseOptions = new()
    {
        CommentHandling = JsonCommentHandling.Disallow,
        AllowTrailingCommas = false
    };

    /// <summary>
    /// 扫描 protocols 根目录（含子目录）下的全部 <c>*.json</c>。
    /// 结果按完整路径序排列，保证同一磁盘状态下扫描顺序确定。
    /// 目录不存在时返回空集合，而不是抛异常——空规则库不是应用级致命错误。
    /// </summary>
    public IReadOnlyList<ProtocolLoadResult> LoadDirectory(string protocolsRoot)
    {
        ArgumentException.ThrowIfNullOrWhiteSpace(protocolsRoot);

        if (!Directory.Exists(protocolsRoot))
        {
            return Array.Empty<ProtocolLoadResult>();
        }

        var paths = Directory
            .EnumerateFiles(protocolsRoot, SearchPattern, SearchOption.AllDirectories)
            .OrderBy(path => path, StringComparer.Ordinal)
            .ToArray();

        var results = new List<ProtocolLoadResult>(paths.Length);
        foreach (var path in paths)
        {
            results.Add(LoadFile(path));
        }

        return results;
    }

    /// <summary>读取并解析单个 Protocol 文件。任何失败都转成 Diagnostic，不向上抛。</summary>
    public ProtocolLoadResult LoadFile(string filePath)
    {
        ArgumentException.ThrowIfNullOrWhiteSpace(filePath);

        string raw;
        try
        {
            raw = File.ReadAllText(filePath);
        }
        catch (Exception ex) when (ex is IOException or UnauthorizedAccessException or SecurityException)
        {
            return ProtocolLoadResult.Failed(new ProtocolDiagnostic
            {
                Code = DiagnosticCode.FileReadFailed,
                Severity = DiagnosticSeverity.Error,
                FilePath = filePath,
                Message = $"无法读取 Protocol 文件：{ex.Message}"
            });
        }

        try
        {
            using var document = JsonDocument.Parse(raw, ParseOptions);
            var identity = ProtocolIdentity.Read(document.RootElement);
            return ProtocolLoadResult.Parsed(
                new ProtocolFile(filePath, Path.GetFileName(filePath), raw, identity));
        }
        catch (JsonException ex)
        {
            return ProtocolLoadResult.Failed(new ProtocolDiagnostic
            {
                Code = DiagnosticCode.JsonParseFailed,
                Severity = DiagnosticSeverity.Error,
                FilePath = filePath,
                JsonPath = ex.Path,
                Message = $"JSON 解析失败（行 {ex.LineNumber}，列 {ex.BytePositionInLine}）：{ex.Message}"
            });
        }
    }
}

File: app/js/protocol-loader.js
/* AI Council v0.1 — D1-R1
 * ProtocolLoader：File 文本 → JSON.parse → Parsed Object。
 * 唯一职责：读取与解析。不做 Schema 校验、不做语义校验、不碰 Runtime / UI。
 */
(function (root) {
  "use strict";

  var D = root.AICouncil && root.AICouncil.Diagnostic;

  /* 解析失败时也要尽量给出可读的身份信息，因此做防御式取值 */
  function readIdentity(parsed) {
    if (!parsed || typeof parsed !== "object" || Array.isArray(parsed)) {
      return { protocolId: null, version: null, name: null, schemaVersion: null };
    }
    return {
      protocolId: typeof parsed.protocol_id === "string" ? parsed.protocol_id : null,
      version: typeof parsed.version === "string" ? parsed.version : null,
      name: typeof parsed.name === "string" ? parsed.name : null,
      schemaVersion: typeof parsed.schema_version === "string" ? parsed.schema_version : null
    };
  }

  function fail(filePath, code, message, details) {
    return Object.freeze({
      ok: false,
      filePath: filePath,
      parsed: null,
      identity: readIdentity(null),
      diagnostics: Object.freeze([
        D.create({ code: code, severity: D.SEVERITY.ERROR, filePath: filePath, jsonPath: "$", message: message, details: details || null })
      ])
    });
  }

  /* 输入：Snapshot 中的 protocolFile 记录 {path, text, readError} */
  function load(record) {
    if (record.readError) {
      return fail(record.path, D.CODE.FILE_READ_FAILED, "无法读取文件：" + record.readError);
    }
    var parsed;
    try {
      parsed = JSON.parse(record.text);
    } catch (e) {
      return fail(record.path, D.CODE.JSON_PARSE_FAILED, "JSON 解析失败：" + (e && e.message ? e.message : String(e)));
    }
    return Object.freeze({
      ok: true,
      filePath: record.path,
      parsed: parsed,
      identity: Object.freeze(readIdentity(parsed)),
      diagnostics: Object.freeze([])
    });
  }

  function loadAll(records) {
    return Object.freeze((records || []).map(load));
  }

  root.AICouncil = root.AICouncil || {};
  root.AICouncil.ProtocolLoader = Object.freeze({ load: load, loadAll: loadAll, readIdentity: readIdentity });
})(typeof globalThis !== "undefined" ? globalThis : this);
